Implement dialog initial focus proposal

josepharhar opened the pull request:
This implements the changes proposed here:
https://github.com/whatwg/html/wiki/dialog--initial-focus,-a-proposal#dialog-draft-text

Specifically:
1. Make the dialog focusing steps look at sequentially focusable elements instead of any focusable element.
2. Make the dialog element itself get focus if it has the `autofocus` attribute set.
3. Make the dialog element itself get focus as a fallback instead of focus being "reset" to the body element.

There are additional open issues around dialog initial focus listed here:
https://github.com/whatwg/html/pull/4184#issuecomment-1072536425

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [ ] At least two implementers are interested (and none opposed):
   * Chrome
   * https://github.com/WebKit/standards-positions/issues/123
   * https://github.com/mozilla/standards-positions/issues/734
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/36831
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1193482
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1811129
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=250795

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/8199/interaction.html" title="Last updated on Jan 20, 2023, 7:17 PM UTC (494d638)">/interaction.html</a>  ( <a href="https://whatpr.org/html/8199/f00c939...494d638/interaction.html" title="Last updated on Jan 20, 2023, 7:17 PM UTC (494d638)">diff</a> )
<a href="https://whatpr.org/html/8199/interactive-elements.html" title="Last updated on Jan 20, 2023, 7:17 PM UTC (494d638)">/interactive-elements.html</a>  ( <a href="https://whatpr.org/html/8199/f00c939...494d638/interactive-elements.html" title="Last updated on Jan 20, 2023, 7:17 PM UTC (494d638)">diff</a> )

domenic commented on source:
```diff
@@ -59158,7 +59158,7 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>] attribute boolean <span data-x="dom-dialog-open">open</span>;
   attribute DOMString <span data-x="dom-dialog-returnValue">returnValue</span>;
-  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-show">show</span>();
+  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-show">show</span>(optional boolean preventInitialFocus);
```
This should be a dictionary, not an optional boolean: see https://w3ctag.github.io/design-principles/#prefer-dict-to-bool

josepharhar replied:
done, thanks!

domenic commented on source:
```diff
@@ -76934,10 +76936,21 @@ END:VCARD</pre>
 
     <ol>
```
These modifications will probably need to be rebased on https://github.com/whatwg/html/pull/8174 , unfortunately. I'm not sure how bad the conflicts will be. It's probably best to change the target branch, or we could wait on this until that gets merged. (Any help checking the test coverage for that PR would be appreciated!)

josepharhar replied:
It looks like that PR has been merged, and this PR is up to date as of 5 days ago

domenic reviewed (commented):
- This does not implement the change where calling `showModal()` on `<dialog autofocus>` will cause the dialog itself to get focus. Probably the best way to spec that is to change the `showModal()` spec to check for the autofocus attribute and, if present, focus the dialog instead of going into the focus delegate steps.

- A big part of the proposal was writing up https://github.com/whatwg/html/wiki/dialog--initial-focus,-a-proposal#improving-conformance-requirements-and-examples as spec text, to replace the current single sentence "The [dialog](https://whatpr.org/html/8199/interactive-elements.html#the-dialog-element) element represents a part of an application that a user interacts with to perform a task, for example a dialog box, inspector, or window."

josepharhar commented:
> A big part of the proposal was writing up https://github.com/whatwg/html/wiki/dialog--initial-focus,-a-proposal#improving-conformance-requirements-and-examples as spec text, to replace the current single sentence "The [dialog](https://whatpr.org/html/8199/interactive-elements.html#the-dialog-element) element represents a part of an application that a user interacts with to perform a task, for example a dialog box, inspector, or window."

Ah thanks, I totally missed this. It is now in the PR

josepharhar commented:
> This does not implement the change where calling showModal() on `<dialog autofocus>` will cause the dialog itself to get focus. Probably the best way to spec that is to change the showModal() spec to check for the autofocus attribute and, if present, focus the dialog instead of going into the focus delegate steps.

I added a commit to do this. Instead of doing it all in showModal I changed the dialog focusing steps because I figured that it would still be important to run the other autofocus related steps at the end of the dialog focusing steps

zcorpan commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <p>Another important aspect of user behavior is whether dialogs are scrollable or not. In some
+   cases, overflow (and thus scrollability) cannot be avoided, e.g., when it is caused by the user's
+   high text zoom settings. But in general, scrollable dialogs are not expected by users, and
+   authors should attempt to avoid them. In particular, authors must not include large blocks of
+   text directly in dialog elements, as this is likely to cause the dialog element itself to
+   overflow. Instead, a scrollable container element should be used, inside the dialog, to contain
```
Don't use "should" here

domenic replied:
E.g., "authors are best off avoiding them" would work.

josepharhar replied:
Done, thanks! I replaced "Instead, a scrollable container element should be used, inside the dialog, to contain this text" with "authors are best off avoiding them"

domenic commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  application, or manually closed by the user.</p>
+
+  <p>Especially for modal dialogs, which are a familiar pattern across all types of applications,
+  authors should work to ensure that dialogs in their web applications behave in a way that is
```
I intended this to be normative when I wrote it. Do you think it's a bad normative requirement, e.g. because it's too vague? I'm not sure on the precedent there.

zcorpan replied:
Sorry for the delay. Yes, I think it's not a good requirement because it's vague and not testable. Further, the wording "authors should work to ensure" makes the requirement about the work the author does, which is a bit off.

If it's not testable, I think making it non-normative is better. e.g. s/should work/are encouraged/

domenic replied:
I don't see any remaining normative text in the section we're talking about, so I think this is resolved?

josepharhar replied:
This paragraph is still there and still says "authors should work to ensure".
Perhaps we could just change it to "authors are encouraged"?

domenic commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  </div>
+
+  <div class="note">
+   <!-- NON-NORMATIVE SECTION -->
```
I think this should not be a note, and doesn't need the "non-normative section" banner. Instead, as @zcorpan says, just avoid using any [magic normative words](https://infra.spec.whatwg.org/#conformance).

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+    &lt;p>By placing an order via this Web site on the first day of the fourth month of the year
+    2010 Anno Domini, you agree to grant Us a non-transferable option to claim, for now and for
+    ever more, your immortal soul.&lt;/p>
+    &lt;p> Should We wish to exercise this option, you agree to surrender your immortal soul,
```
```suggestion
    &lt;p>Should We wish to exercise this option, you agree to surrender your immortal soul,
```

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  &lt;/form>
+&lt;/dialog></code></pre>
+
+   <p>Note how the dialog focusing steps would have picked the scrollable div element by default,
```
Make sure to link all element and attribute names in these paragraphs (autofocus, div, p)

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <p>In contrast, if the p elements expressing the terms of service did not have such a wrapper div
+   element, then the dialog itself would become scrollable, violating the above conformance
+   requirements. Furthermore, in the absence of any autofocus attribute, such a non-conformant
+   markup pattern would have tripped up the dialog focusing steps' default behavior, and caused
```
```suggestion
   markup pattern would have tripped up the <span>dialog focusing steps</span>'s default behavior, and caused
```

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59243,8 +59337,9 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
-  <p>When the <dfn method for="HTMLDialogElement"><code data-x="dom-dialog-show">show()</code></dfn>
-  method is invoked, the user agent must run the following steps:</p>
+  <p>When the <dfn method for="HTMLDialogElement"><code
+  data-x="dom-dialog-show">show(dialogShowOptions)</code></dfn> method is invoked, the user agent
```
```suggestion
  data-x="dom-dialog-show">show(<var>options</var>)</code></dfn> method is invoked, the user agent
```

to match the IDL.

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59256,7 +59351,8 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
    <span>focused</span> element.</p></li>
 
-   <li><p>Run the <span>dialog focusing steps</span> for the <code>dialog</code> element.</p></li>
+   <li><p>If <var>dialogShowOptions</var>'s preventInitialFocus flag is false, run the <span>dialog
```
```suggestion
   <li><p>If <var>dialogShowOptions</var>["<dfn dict-member for="DialogShowOptions"><code data-x="dom-dialogshowoptions-preventinitialfocus">preventInitialFocus</code></dfn>"] is false, then run the <span>dialog
```

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+};
+
+dictionary DialogShowOptions {
+  boolean preventInitialFocus = false;
```
```suggestion
  boolean <span data-x="dom-dialogshowoptions-preventinitialfocus">preventInitialFocus</span> = false;
```

(only works when combined with one of my below changes)

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59256,7 +59351,8 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
 
-   <li><p>Run the <span>dialog focusing steps</span> for the <code>dialog</code> element.</p></li>
+   <li><p>If <var>dialogShowOptions</var>'s preventInitialFocus flag is false, run the <span>dialog
+   focusing steps</span> for the <code>dialog</code> element with is modal set to false.</p></li>
```
```suggestion
   focusing steps</span> given the <code>dialog</code> element and false.</p></li>
```

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59298,13 +59394,22 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
    <span>focused</span> element.</p></li>
 
-   <li><p>Run the <span>dialog focusing steps</span> for <var>subject</var>.</p></li>
+   <li><p>Run the <span>dialog focusing steps</span> for <var>subject</var> with is modal set to
+   true.</p></li>
```
```suggestion
   <li><p>Run the <span>dialog focusing steps</span> given <var>subject</var> and true.</p></li>
```

josepharhar replied:
done

josepharhar commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  familiar to users of non-web applications.</p>
+
+  <p class="XXX">The semantics and suggested usage of modeless dialogs (i.e., those shown using
+  show() instead of showModal()) are unclear, and we are even considering deprecating the show()
```
done, thanks!

josepharhar commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
   attribute DOMString <span data-x="dom-dialog-returnValue">returnValue</span>;
-  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-show">show</span>();
+  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-show">show</span>(optional
+  DialogShowOptions options = {});
```
done, thanks!

josepharhar commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>] undefined <span data-x="dom-dialog-close">close</span>(optional DOMString returnValue);
+};
+
+dictionary DialogShowOptions {
```
done

josepharhar commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   used autofocus to override the default.</p>
+
+   <p>Even if the author wants to focus the Product Number field by default, they are best off
+   explicitly specifying that by using autofocus on that input element. This makes the intent
```
done

josepharhar commented on source:
```diff
@@ -59158,16 +59158,110 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+
+   <p>In contrast, if the p elements expressing the terms of service did not have such a wrapper div
+   element, then the dialog itself would become scrollable, violating the above conformance
+   requirements. Furthermore, in the absence of any autofocus attribute, such a non-conformant
```
done

josepharhar commented:
Thanks for the review @domenic! I believe I've addressed all your comments

domenic commented on source:
```diff
@@ -59155,16 +59155,106 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  dialog element itself to overflow. Authors are best off avoiding them.</p>
+
+  <div class="example">
+   The following terms of service dialog respects the above requirements.
```
```suggestion
   <p>The following terms of service dialog respects the above suggestions.</p>
```

(We should avoid "requirements" now that we have stopped using "should" in the above paragraph.)

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59155,16 +59155,106 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+
+   <p>In contrast, if the <code>p</code> elements expressing the terms of service did not have such
+   a wrapper <code>div</code> element, then the <code>dialog</code> itself would become scrollable,
+   violating the above conformance requirements. Furthermore, in the absence of any <code
```
Since we 
```suggestion
   violating the above advice. Furthermore, in the absence of any <code
```

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59155,16 +59155,106 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   a wrapper <code>div</code> element, then the <code>dialog</code> itself would become scrollable,
+   violating the above conformance requirements. Furthermore, in the absence of any <code
+   data-x="attr-fe-autofocus">autofocus</code> attribute, such a markup pattern would have violated
+   the above advice and tripped up the <span>dialog focusing steps</span>' default behavior, and
```
```suggestion
   the above advice and tripped up the <span>dialog focusing steps</span>'s default behavior, and
```

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59295,13 +59389,21 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
 
-  <p>The <dfn>dialog focusing steps</dfn> for a <code>dialog</code> element <var>subject</var> are as follows:</p>
+  <p>The <dfn>dialog focusing steps</dfn>, given a <code>dialog</code> element <var>subject</var>
+  and a boolean <var>is modal</var>, are as follows:</p>
```
Let's use isModal; more modern algorithms prefer camelCase.

josepharhar replied:
done

domenic commented on source:
```diff
@@ -59295,13 +59389,21 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p>If <var>control</var> is null, then set <var>control</var> to the <span>focus
+   delegate</span> of <var>subject</var>.</p></li>
 
    <li><p>If <var>control</var> is null, then set <var>control</var> to <var>subject</var>.</p></li>
```
We need to try to make the dialog itself focusable. Since the spec never actually mandates focusability of any given element, this actually consists of updating some hints scattered through the spec:

- Adding it to the first row of example elements in the table https://whatpr.org/html/8199/interaction.html#dom-anchor , which currently reads "`iframe, <input type=text>, sometimes <a href="">`. This isn't actually a guarantee, because the spec never actually requires UAs to make anything in particular focusable. But it's a recommendation.

-  Update the note in step https://whatpr.org/html/8199/interactive-elements.html#dialog-focusing-steps from "For modal dialogs, this means that any [earlier modifications](https://whatpr.org/html/8199/interactive-elements.html#note-dialog-plus-focus-fixup) to the [focused area of the document](https://whatpr.org/html/8199/interaction.html#focused-area-of-the-document) will apply." to something like "This would only happen if _subject_ had no focus delegate, and the user agent decided that `dialog` elements were not generally focusable. In that case, any earlier modifications..."

domenic commented on source:
```diff
@@ -59325,22 +59412,32 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
-    modal dialogs, this means that any <a href="#note-dialog-plus-focus-fixup">earlier
-    modifications</a> to the <span>focused area of the document</span> will apply.</p>
+    <p class="note">If <var>control</var> is not <span>focusable</span>, this will do nothing. This
+    would only happen if subject had no focus delegate, and the user agent decided that dialog
```
```suggestion
    would only happen if subject had no focus delegate, and the user agent decided that <code>dialog</code>
```

domenic commented on source:
```diff
@@ -76246,7 +76343,9 @@ partial interface <span id="NavigatorUserActivation-partial">Navigator</span> {
      <td headers="td-fa-1 th-fa-examples" colspan=2>
      <p class="example"><code>iframe</code>, <code data-x="attr-input-type-text">&lt;input
-     type=text></code>, sometimes <code data-x="a">&lt;a href=""></code> (depending on platform
+     type=text></code>, <code data-x="dialog">&lt;dialog></code>, sometimes <code data-x="a">&lt;a
```
Let's do this as `iframe, dialog, <input type=text>, sometimes <a href="">`. Basically use bare tag names first, and then go into the ones that are attribute dependent, and use angle brackets for those.

josepharhar replied:
done

domenic reviewed (approved):
LGTM with two nits plus the block on https://github.com/whatwg/html/pull/8174. Once that is merged I can try to take care of the rebasing.

I guess we also need tests...

domenic commented:
OK, this is looking good! #8174 got merged and I merged the changes into here, so this is good to go from my perspective.

Can you update the original post with a full description of the changes? It's a bit outdated, still mentioning preventInitialFocus and a TODO that I believe is taken care of. And it doesn't mention all the improvements to the explanatory text, etc.

At this point I believe we're waiting on tests mainly, since there was high-level consensus on these changes from all browsers some time ago. Still, now that we have a clear diff, review appreciated from @emilio @annevk @nt1m for the processing model changes, and if @zcorpan wants to do another review, or @sideshowbarker wants to do his review, on the conformance requirements / explanatory text, that would be appreciated.

annevk commented on source:
```diff
@@ -59218,8 +59218,94 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
-  to perform a task, for example a dialog box, inspector, or window.</p>
+  <p>The <code>dialog</code> element represents a transitory part of an application, in the form of
+  a small window ("dialog box"), which the user interacts with to perform a task or gather
+  information. Once the user interaction is complete, the dialog can be automatically closed by the
```
s/interaction is complete/is done/

"User interaction" is quite overloaded already.

josepharhar replied:
done

annevk commented on source:
```diff
@@ -59218,8 +59218,94 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  dialog is shown, but might not be a substitute for authors carefully thinking through the correct
+  choice to match user expectations for a specific dialog. As such, authors should use the <code
+  data-x="attr-fe-autofocus">autofocus</code> attribute on the descendant element of the dialog that
+  the user will expect to immediately interact with after the dialog opens. If there is no such
```
s/will expect/is expected/

josepharhar replied:
done

annevk commented on source:
```diff
@@ -59218,8 +59218,94 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   explicitly specifying that by using autofocus on that <code>input</code> element. This makes the
+   intent obvious to future readers of the code, and ensures the code stays robust in the face of
+   future updates. (For example, if another developer added a close button, and positioned it in the
+   DOM before the Product Number field).</p>
```
s/DOM/node tree/

josepharhar replied:
done

annevk commented on source:
```diff
@@ -59218,8 +59218,94 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   explicitly specifying that by using autofocus on that <code>input</code> element. This makes the
+   intent obvious to future readers of the code, and ensures the code stays robust in the face of
+   future updates. (For example, if another developer added a close button, and positioned it in the
+   DOM before the Product Number field).</p>
```
Reading this example it seems a little weird to me we'd autofocus readonly controls, but the majority of implementations does seem to be doing that. (Firefox does not though.)

josepharhar replied:
I see.
What if I replaced readonly with value=foo and said that the second input should be focused because the first one is pre-filled...?
I'm having a hard time thinking of another example because I don't know anything else which probably shouldn't be focused by the browser but actually is.

annevk replied:
Not sure either. I suppose it could also be some kind of button along the lines of "add product" or some such.

josepharhar replied:
ok, i replaced it with an "add product" button and changed the explanation

annevk commented on source:
```diff
@@ -76884,11 +76981,18 @@ partial interface <span id="NavigatorUserActivation-partial">Navigator</span> {
+     <var>descendant</var>.</p></li>
+
+     <li><p>Otherwise, if <var>focusTarget</var> is not a <code>dialog</code> and
+     <var>descendant</var> is a <span>focusable area</span>, then set <var>focusableArea</var> to
```
Nit: we typically don't use "then" in an "Otherwise" clause. Only for lone "if" clauses.

josepharhar replied:
done

annevk reviewed (commented):
Looks good editorially, found a couple nits.

annevk commented on source:
```diff
@@ -59219,8 +59219,94 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  <p class="XXX">The semantics and suggested usage of modeless dialogs (i.e., those shown using
+  <code data-x="dom-dialog-show">show()</code> instead of <code
+  data-x="dom-dialog-showModal">showModal()</code>) are unclear, and we are even considering
+  deprecating the <code data-x="dom-dialog-show">show()</code> method entirely.  Please join the
```
```suggestion
  deprecating the <code data-x="dom-dialog-show">show()</code> method entirely. Please join the
```

josepharhar replied:
done

annevk reviewed (approved):
Thanks! Looks good modulo tiny nit. Still not entirely happy about the example, but cannot think of something better and this at least avoids the readonly problem.

josepharhar commented:
I finally got a draft of the implementation in chromium done, and here is a new test and many other test changes to account for the focus ring being rendered on dialogs: https://github.com/web-platform-tests/wpt/pull/36498
I tried my best to keep the tests affected tests passing with or without this new behavior, but a couple of them will fail on firefox and safari and are probably tracked in interop22...

nt1m commented on source:
```diff
@@ -59219,8 +59219,94 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  authors should work to ensure that dialogs in their web applications behave in a way that is
+  familiar to users of non-web applications.</p>
+
+  <p class="XXX">The semantics and suggested usage of modeless dialogs (i.e., those shown using
+  <code data-x="dom-dialog-show">show()</code> instead of <code
+  data-x="dom-dialog-showModal">showModal()</code>) are unclear, and we are even considering
+  deprecating the <code data-x="dom-dialog-show">show()</code> method entirely. Please join the
+  discussion in <a href="https://github.com/whatwg/html/issues/7707">issue #7707</a>.</p>
```
Can we remove this note? I think it's too early to say at this point that the popup API will replace `dialog.show()`. cc @annevk .

josepharhar replied:
This doesn't explicitly talk about popup/popover. However, I don't have strong feelings about including or excluding it.

annevk replied:
Can you please remove it for now then?

josepharhar commented:
> This does not implement the change where calling showModal() on `<dialog autofocus>` will cause the dialog itself to get focus.

Does this mean that non-modal dialogs should not be focusable?

Implementing this change in chromium has resulted in open modal dialogs preventing the body from being focused when clicking outside of the dialog - that just focuses the dialog. Is that ok...?

domenic commented:
> Does this mean that non-modal dialogs should not be focusable?

I don't think we've given a lot of thought to non-modal dialogs. Good question. I would tend toward them not being focusable...

> Implementing this change in chromium has resulted in open modal dialogs preventing the body from being focused when clicking outside of the dialog - that just focuses the dialog. Is that ok...?

This is surprising, and probably not desired? It might be good to consult the original folks behind the proposal to be sure, but I at least expect that clicking outside should cause the body to be focused.

Probably this is happening because technically the `::backdrop` is part of the dialog, and so clicking on it focuses the dialog. Maybe if you add `pointer-events: none !important` like you did for `popover=""`??

josepharhar commented:
> Maybe if you add pointer-events: none !important like you did for popover=""??

That did the trick! I've added it to this PR

> I don't think we've given a lot of thought to non-modal dialogs. Good question. I would tend toward them not being focusable...

I wonder if @scottaohara has any thoughts. I don't think there was any mention of different logic between them in the proposal/wiki he wrote:

> Otherwise, if nothing inside the dialog is tabbable, focus the dialog itself, so that at least something gets focused.

josepharhar commented:
Another side effect of my current chromium implementation is that it makes the dialog element keyboard focusable, meaning that I can focus the dialog element by pressing shift+tab when the first focusable element of the dialog is focused. Is this desired behavior...?

I talked to Mason and he mentioned that we could also try to make it only keyboard focusable if it has no keyboard focusable descendants.

Any thoughts? @scottaohara

scottaohara commented:
The intent here was not to have the dialog element be included in the tab-key focus order for the web page. So if there's a focusable element inside of the dialog, someone should not be able to focus/re-focus the dialog, even if the dialog was focused by default.   But the use case of a modal dialog being rendered without any focusable element within could be awkward for some people who are trying to enter a web page from the browser chrome by use of the tab key, but then finding that they cannot.  Someone 'should', at least on windows, be able to use the F6 key to still move to the web document in that particular case.  


re: body not being focused when a modal dialog is open, body and its contents should be inert, so why would focus be expected to go there?  Unless I'm missing something, mouse click shouldn't be able to focus the inert content, so it'd make sense for the backdrop to block that.  Otherwise, the content outside of the modal dialog is not fully inert?  Again, am I missing something?  The pointer events none was added for popover because a popover is not supposed to be modal / cause content outside of it to be inert, and the mouse click to the body was to trigger light dismiss behavior.


re: focusing differences between modal / non-modal dialogs.  I'm not sure why they'd be different unless focus was not moved to a non-modal dialog when it was invoked.  But if focus moves to it, and there's the use case where there's no focusable element within... then where would focus go?  If it remained on the element (likely button) that invoked the non-modal dialog, then that seems reasonable to me.  If someone used an `autofocus` attribute on the non-modal dialog, then i'd still expect it to go to the dialog element itself.  Otherwise there'd need to be a stipulation that autofocus can be used on a modal dialog, but not on a non-modal, but the only way to declare a dialog one way or the other is with the show or showModal methods.  so outside of the attribute just failing to work, there wouldn't be a good way to surface this to the author through a markup validation error, unless i'm not thinking of something?

josepharhar commented:
> The intent here was not to have the dialog element be included in the tab-key focus order for the web page

Say no more, I'll implement it this way!

> re: body not being focused when a modal dialog is open, body and its contents should be inert, so why would focus be expected to go there? Unless I'm missing something, mouse click shouldn't be able to focus the inert content, so it'd make sense for the backdrop to block that. Otherwise, the content outside of the modal dialog is not fully inert? Again, am I missing something? The pointer events none was added for popover because a popover is not supposed to be modal / cause content outside of it to be inert, and the mouse click to the body was to trigger light dismiss behavior.

This makes sense, but the current behavior in all browsers is that clicking outside of the dialog element focuses the body element, not the dialog. I can change it but I feel like it's another risk of breaking stuff and doesn't seem like the main point of this change. If you insist though I'll happily go for it.

josepharhar commented:
> > Implementing this change in chromium has resulted in open modal dialogs preventing the body from being focused when clicking outside of the dialog - that just focuses the dialog. Is that ok...?
> 
> This is surprising, and probably not desired? It might be good to consult the original folks behind the proposal to be sure, but I at least expect that clicking outside should cause the body to be focused.
> 
> Probably this is happening because technically the ::backdrop is part of the dialog, and so clicking on it focuses the dialog. Maybe if you add pointer-events: none !important like you did for popover=""??

It turns out that this makes a WPT fail: https://github.com/web-platform-tests/wpt/pull/31758

@nt1m I'm guessing that you don't think we should add `pointer-events: none` to `::backdrop`. What was the rationale?

> re: body not being focused when a modal dialog is open, body and its contents should be inert, so why would focus be expected to go there? Unless I'm missing something, mouse click shouldn't be able to focus the inert content, so it'd make sense for the backdrop to block that. Otherwise, the content outside of the modal dialog is not fully inert? Again, am I missing something? The pointer events none was added for popover because a popover is not supposed to be modal / cause content outside of it to be inert, and the mouse click to the body was to trigger light dismiss behavior.

I guess that maybe the body never getting focus is OK... I think that made something tricky happen with the implementation or tests but it's been a while, I can give it a shot and see what happens.

nt1m commented:
@josepharhar That test was added for https://whistlr.info/2021/in-defence-of-dialog/#sidebar-%2F-nav-bar . See the part that says it works in Chrome, but not Safari.

Basically Safari used to consider the ::backdrop not as part of the dialog when querying the event target when clicking the ::backdrop, unlike Chrome & Firefox which considered ::backdrop as part of the element.

I don't have strong opinions on how this should behave fwiw, it was just a test to ensure interop.

annevk commented:
There are still a couple of unresolved comment threads. It appears this also needs rebasing.

josepharhar commented:
> There are still a couple of unresolved comment threads. It appears this also needs rebasing.

I have resolved the comment threads and merged with a commit from 5 days ago.

> > re: body not being focused when a modal dialog is open, body and its contents should be inert, so why would focus be expected to go there? Unless I'm missing something, mouse click shouldn't be able to focus the inert content, so it'd make sense for the backdrop to block that. Otherwise, the content outside of the modal dialog is not fully inert? Again, am I missing something? The pointer events none was added for popover because a popover is not supposed to be modal / cause content outside of it to be inert, and the mouse click to the body was to trigger light dismiss behavior.
> 
> I guess that maybe the body never getting focus is OK... I think that made something tricky happen with the implementation or tests but it's been a while, I can give it a shot and see what happens.

> That test was added for https://whistlr.info/2021/in-defence-of-dialog/#sidebar-%2F-nav-bar . See the part that says it works in Chrome, but not Safari.

I removed the pointer-events change and now the test is passing again.

I am happy with this being merged whenever ready. I am going to ship the changes in chromium soon.
It looks like the last build failure is due to some infrastructure rather than my patch, it builds locally on my machine.

domenic commented on source:
```diff
@@ -59129,8 +59129,88 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <p>In the following example, a dialog is used for editing the details of a product in an
+   inventory management web application.</p>
+
+   <pre><code class="html">&lt;dialog>
```
I'm sorry, this is totally on me, but I forgot why we changed this from the example at https://github.com/whatwg/html/wiki/dialog--initial-focus,-a-proposal#example . @scottaohara I vaguely remember you being involved?

I'm not very happy with this example because it's just hard to believe the web developer would put the button before the input. The example at https://github.com/whatwg/html/wiki/dialog--initial-focus,-a-proposal#example is more believable. If there's something wrong with my initial version (I vaguely recall there was), could we try to come up with a more believable example here?

domenic replied:
Ah, I found the conversation. It's at https://github.com/whatwg/html/pull/8199#discussion_r1002961552 (collapsed by default). I guess @annevk didn't like the example because it relied on the focusablity of readonly controls? I don't think that's so strange. The point of readonly (as opposed to disabled) is explicitly to be focusable/selectable. E.g. it's what I use in the second textbox in https://domenic.github.io/rewrapper/ .

@annevk , are you OK reverting this to the previous example?

scottaohara replied:
agreed @domenic, far prefer the original example as this new one doesn't make sense.  the original example is something i can point to many instances of in microsoft products. 

i'm confused by the original comment, indicating that Firefox did not autofocus readonly controls.  [testing today](https://codepen.io/scottohara/full/WNKdOKL), focus goes to the readonly text field just fine (chrome and firefox), as i'd expect.  So maybe Firefox has reversed course to match up with other implementations since october?

josepharhar replied:
Ok I tentatively reverted back to the previous example

annevk replied:
Alright, fair enough.

domenic reviewed (approved):
Normative parts LGTM. I have a quibble about the example, but everyone should consider this ready to implement.

It would be good to get explicit reaffirmation from other implementers that they are still on-board with this plan. I guess we can tag this with agenda+, and remove that if they respond to the standards positions request before the next meeting.

annevk reviewed (approved):
I'm okay with merging this.
